[ASM] - Expander - Service Ownership Model Update

kball-pa opened the pull request:
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-hq.paloaltonetworks.local/browse/EXPANDR-4989

## Description
Updates the Service Ownership playbook to use a more sophisticated machine learning model. In particular, the RankServiceOwners script is updated to:

- download the model from GCS and load it
- turn the data in `asmserviceownerunrankedraw` into a numerical array of features
- run inference on the featurized set of potential owners
- normalize this score to a value between 0.5 and 1
- rank them according to this normalized score

## Must have
- [ x] Tests
- [ x] Documentation

kball-pa commented:
@anas-yousef @efelmandar are the security issues flagged above (using dill for serialization/de-serialization) blocking issues? for context this automation reads a pickled machine learning model from a remote GCS bucket with privileged write access

kball-pa commented:
Converted PR to draft to block merging until testing with final prod GCS bucket is done.

Testing is currently blocked on devops work to grant read access to the relevant service accounts.

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -202,16 +299,300 @@ def _get_k(
+        """
+        Initialize a featurization pipeline.
+
+        As of May 2023, we use `sources` to parse out which sources are included in the
```
If I'm reading it right, this comment is not a comment on the  OwnerFeaturizationPipeline, so much as a process recommendation -- that we should flag for all the relevant teams that by using the "source" property on the `owners` provided upstream, we are coupling this code & model releases tightly to the actual values in that field. And if we can make a better recommendation, we should do so.

So, could you:
1. socialize the flag/concern/recommendation from this paragraph with the relevant teams, and provide a suggestion for a better implementation of the input `owners` format, and figure out who will make the change and by which release, or whether we won't prioritize it at all.
2. make sure there is a note about "if you change the source values, you'll need to update the ML service ownership model". in the document that that team is maintaining around the downstream implications of hardcoded values (I believe @johnnywilkes owns this document re: the other ongoing initiative, and can direct you to it)
3. remove this comment from this PR (because it isn't about using the code at all)

kball-pa replied:
@BigEasyJ and I discussed the addition of field in `asmserviceowner` for this specific info; ticketed it and John will see if he can get it prioritized for the 2.3 Expander release

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -202,16 +299,300 @@ def _get_k(
+            # `Packs/CortexAttackSurfaceManagement`
+            # for `Name,Email,Source,Timestamp`.
+
+            # FIXME (plt 2023.06): work with RC to retain Source System directly from the `owners` metadata
```
@kball-pa let's try to resolve this FIXME before merge.

ideally it looks like we'd have RC add a field upstream that is specifically for the "source system" concept. can you look into whether this is feasible on a reasonable timeframe?

kball-pa replied:
see https://github.com/demisto/content/pull/28375#discussion_r1276594560

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -202,16 +299,300 @@ def _get_k(
+        for sample_idx, owner in enumerate(owners):
+            for feature_idx, (method_name, method) in enumerate(self.FEATURES):
+                try:
+                    if "similarity" in method_name and "person" in method_name and "asset" in method_name:
+                        X[sample_idx, feature_idx] = method(service_identifiers, owner)
+                    else:
+                        X[sample_idx, feature_idx] = method(owner)
```
a lot of the stuff that I originally wrote reads like unreviewed prototype code (because it is), but this REALLY reads like unreviewed prototype code -- is there a more robust way to handle it? maybe instead of a single variable `self.FEATURES` there should be 2 variables? other ideas?

(feel free to fix & review the other code too!)

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -1,87 +1,64 @@
+import dill as pickle
+
+
+def test_load_pickled_xpanse_object():
```
this test is dangerous -- if something exceptional happens, we'll end up with the model on disk being `[1,2,3]`, and it will exist and have non-zero size, so all featurizations afterwards will always fail for that tenant, until the disk is somehow cleared.

instead, I'd recommend:
* `load_pickled_xpanse_object` should take the path as an input. the path should NOT be a public constant, and it should not be imported in this testing file
* in the test, you should use Python's `tempfile` library to manage the temporary existence of this file -- it is careful about cleanup under extraordinary conditions, and it clearly marks what files won't persist

and/or don't test this code at all.

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -1,87 +1,64 @@
 
 @pytest.mark.parametrize('owners,expected_out', [
     (
         # returned in sorted order
```
I'm struggling to understand what's being tested here. is the idea that we're passing in arbitrary numbers and they're getting returned in sorted order? if so, given these inputs & outputs, it seems simple -- why is it worth a method in our codebase for this instead of relying on `sorted` (or `list.sort`)? and if sorting is what's going on and it *is* more complex than can be reasonably handled by a passsthrough to `sort`, could you rename the method to be clearer? (maybe capturing that it's a `sort` more than a `rank`, or naming with another characteristic that matters in making a plain `sorted` not appropriate?)

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -199,55 +176,135 @@ def test_aggregate(owners, expected_out):
+    Test the default scoring case
+    """
+    model_mock = Mock()
+    model_mock.predict.return_value = np.array([4, 7, 31, 0.33, 15])
```
where did these numbers come from? and the ones in line 215?

If the right value of what we are testing isn't readily apparent to the reader of the test, we should change something about the test -- tests should be verifying obvious functionality. some options might be:
1. break apart the test so the source of the numbers is apparent (maybe into multiple tests -- e.g., maybe we test featurization of the `owners` values separate from model prediction)
2. add enough context so the number source is apparent
3. make assertions about something that *is* apparent from the data (like who is the most / least likely to be the best owner point-of-contact, or that we do get a numeric answer)

you've already got 1, 2, and 3 covered below -- so after reviewing the rest of this doc, I'm wondering, does this test cover something we aren't testing elsewhere? could we simplify it?

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -525,3 +617,198 @@ def test_get_k_bad_values(target_k, k_tol, a_tol, min_score_proportion, expected
+    # has 1/2 matched
+    pipeline = OwnerFeaturizationPipeline()
+    out = pipeline.get_name_similarity_person_asset(["amira-instance", "abc-123"], owner)
+    assert out > 1
```
should this be >= 1, since there is only one obvious match?

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -199,55 +176,135 @@ def test_aggregate(owners, expected_out):
+    assert np.allclose(scores_out, [0.53, 0.56, 0.77, 0.5, 0.63])
+
+
+def test_score_model_load_fail(mocker):
```
can this be simplified? the `owners` seem very complex if you just want to verify that if `load_pickled_xpanse_object` raises an exception, we still get an output.

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -199,55 +176,135 @@ def test_aggregate(owners, expected_out):
+    assert np.allclose(scores_out, [SCORE_LOWER_BOUND for _ in range(len(owners))])
+
+
+def test_score_model_inference_fail(mocker, caplog):
```
can this be simplified? the `owners` seem very complex if you just want to verify that if `load_pickled_xpanse_object` raises an exception, we still get an output.

ptoman-pa replied:
should `test_score_model_inference_fail` be the same as `test_score_model_load_fail`? they look very similar to me -- if they're different, and if even after simplifying they still look very similar, could you call out the difference in what you're trying to test in the docstring?

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -1,87 +1,64 @@
 import demistomock as demisto  # noqa: F401
```
There's a docstring format for tests that I've had to use in this codebase before -- something about When/Given/Then or the like. you'll probably be asked to refactor to this format (you can find the format in a parallel test file) -- it might be worth confirming this and then getting ahead of it.

BigEasyJ replied:
+1 Please you docstrings for all functions.

ptoman-pa reviewed (commented):
big PR -- lots of progress!

kball-pa commented on Packs/CortexAttackSurfaceManagement/.pack-ignore:
```diff
@@ -41,3 +41,4 @@ ReportHeader
 RCS
 SNMP
 NMAP
+ml
```
yeah it's an allow list. not sure about space tokenization; in this case it used dash tokenization -- the release-notes check failed on the docker image tag "xpanse-ml-ev2" (with an added message like "did you mean 'xml'"?) and prompted me to add "ml" to the .pack-ignore. 

I guess i could add the full image tag "xpanse-ml-ev2" to the ignore list?

kball-pa replied:
after discussing with pamela: planning to leave as-is (unless others recommend otherwise)

ptoman-pa replied:
leaving alone makes sense to me -- adding "ml" to the spellchecker seems very reasonable.

kball-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/GetProjectOwners/GetProjectOwners.py:
```diff
@@ -30,7 +30,7 @@ def extract_project_name(service_account: str) -> str:
 
 
-def get_iam_policy(project_name: str) -> List[Dict[str, Any]]:
+def get_iam_policy(project_name: str) -> list[dict[str, Any]]:
```
might be a new demisto standard? with the imported type hints from `typing` the `pre-commit` check in CI/CD failed and outputted a diff removing the imported type hints in favor of the built-in types, and once i applied the diff the validation pass. but that didn't happen on my previous PR's

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md:
```diff
@@ -0,0 +1,18 @@
+
+- Updated the playbook to pass ASM system IDs as input to the RankServiceOwners automation.
+
+#### Scripts
```
It depends. If something would be relevant for a customer or external parties, yes please put why. If it's more relevant internally, we should use internal options or doc strings / comments in the code for "why"?

ptoman-pa replied:
I think this means "I'm comfortable with these release notes" -- but if you aren't please say so

kball-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -202,16 +299,300 @@ def _get_k(
+
+
+def get_name_similarity_index(
+    personal_monikers: Iterable[str],
```
couldn't come up with anything better but did add an example!

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md:
Please undo this change.

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_27.md:
```diff
@@ -0,0 +1,18 @@
+
+##### Cortex ASM - Service Ownership
+
+- Updated the playbook to pass ASM system IDs as input to the RankServiceOwners automation.
```
```suggestion
Updated the playbook to pass ASM system IDs as input to the RankServiceOwners automation.
```

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:
```diff
@@ -1,92 +1,37 @@
 import unittest
-from RankServiceOwners import score, main, rank, _canonicalize, aggregate, _get_k
+from RankServiceOwners import (
+    score,
```
NIT: Please alphabetize these.

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -4,28 +4,131 @@
-    if owners:
-        max_count = max(owner.get('Count', 1) for owner in owners)
+    REMOTE_GCS_BUCKET = "xpanse-service-ownership-ml-models-dev"
+    REMOTE_GCS_PATH = ""  # ok for this to be empty string
```
Is the a mutable variable? If so I would lowercase it.

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -4,28 +4,131 @@
+        remote_path = posixpath.join(REMOTE_GCS_PATH, file_name)
+
+        demisto.info(f"Starting download of '{file_name}' from gs://{REMOTE_GCS_BUCKET}/{remote_path}")
+        client = google.cloud.storage.client.Client()
```
[Does this work without credentials?](https://cloud.google.com/python/docs/reference/storage/latest/google.cloud.storage.client.Client)

ptoman-pa replied:
^ great catch -- there should be a try/except around the appropriate code, with a meaningful log message if service ownership can't run (it may be all the way out in `main` or otherwise near the beginning of all this functionality)

kball-pa replied:
yes, the invocation is surrounded by a try/except which will assign a fixed fallback score per owner if the model can't be downloaded for any reason

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -202,16 +318,317 @@ def _get_k(
+        Initialize a featurization pipeline.
+        """
+        if sources is None:
+            self.SOURCES = ["Azure", "GCP", "AWS", "Tenable", "Rapid7", "Qualys", "SNOW-CMDB", "Splunk", "PrismaCloud"]
```
After meeting we decided this is the best option for now, especially given the change we made to include [`Owner-In-Tags-From-<PLATFORM>` as a source in `asmserviceowner`](https://github.com/demisto/content/blob/master/Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Enrichment.yml#L1208). There is not an easy way to currently pull a unique list of integrations the service owner was found in.

cc: @johnnywilkes @capanw

BigEasyJ reviewed (changes requested):
Forgot to hit submit yesterday on my review, but I have added additional things since then as well.

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:
```diff
@@ -53,6 +53,40 @@ tasks:
     scriptarguments:
       owners:
         simple: ${alert.asmserviceownerunrankedraw}
+      asmsystemids:
```
I'm assuming this comment means "it's ok to leave this PR as it is" 👍

kball-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -4,28 +4,131 @@
+    # create the cache_path but the file will be empty
+    if not (os.path.exists(cached_file_path) and os.path.getsize(cached_file_path)):
+        remote_path = posixpath.join(REMOTE_GCS_PATH, file_name)
+
```
"infrastructure-related service account"

kball-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -202,16 +299,300 @@ def _get_k(
     return k
 
 
+# Begin: Model Featurization Code
```
i think it's fine!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=28375) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=28375) before we can accept your contribution.<br/>**18** out of **21** committers have signed the CLA.<br/><br/>:white_check_mark: ilaner<br/>:white_check_mark: samuelFain<br/>:white_check_mark: yaakovpraisler<br/>:white_check_mark: omerKarkKatz<br/>:white_check_mark: mmhw<br/>:white_check_mark: content-bot<br/>:white_check_mark: sapirshuker<br/>:white_check_mark: israelpoli<br/>:white_check_mark: DinaMeylakh<br/>:white_check_mark: michal-dagan<br/>:white_check_mark: idovandijk<br/>:white_check_mark: maimorag<br/>:white_check_mark: BEAdi<br/>:white_check_mark: YuvHayun<br/>:white_check_mark: eyalpalo<br/>:white_check_mark: ShahafBenYakir<br/>:white_check_mark: eepstain<br/>:white_check_mark: edik24<br/>:x: kobymeir<br/>:x: tomer-pan<br/>:x: nkanon<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=28375) it.</sub>

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/pack_metadata.json:
Please add the release notes `.md` file.

omerKarkKatz commented:
Hi @kball-pa, @johnnywilkes, we haven’t heard from you in a while.
What is the status here? 
Do you need any help with the pull request?
Please feel free to reach out to me here or on Slack.
Thanks again for contributing to our repo, hope to hear from you soon.

johnnywilkes commented:
@omerKarkKatz , we are still working on this.  We will approve within our group and let you know.

kball-pa commented:
Thanks @johnnywilkes . I've addressed the comments from @BigEasyJ but don't understand the current failures in pre-commit, review_release_notes, and run validations. Lmk if folks see further changes I should make.

kball-pa commented on Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:
```diff
@@ -53,6 +53,55 @@ tasks:
+              field:
+                value:
+                  simple: id
+          filters:
```
@BigEasyJ don't we actually want to remove lines 74-104?

kball-pa replied:
i went ahead and made that change; please review [this commit](https://github.com/demisto/content/pull/28375/commits/f050426e4f9ac760ceeaae503db74b934def8c72)

BigEasyJ commented:
There should only be errors on `pull-check` and `release_notes_review` now and those are known issues that you can ignore.

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:
```diff
@@ -71,37 +71,6 @@ tasks:
               field:
                 value:
                   simple: id
-          filters:
```
Does your script still not need to check for `alert.asmsystemids.type` have the value `ASSET-NAME` or `ASSET-ID`?

BigEasyJ replied:
Discussed: `ASSET-ID` is not needed. `ASSET-NAME` is used for the scripts business logic optionally.

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:
```diff
@@ -53,6 +53,24 @@ tasks:
     scriptarguments:
       owners:
         simple: ${alert.asmserviceownerunrankedraw}
```
```suggestion
        complex:
          root: alert.asmserviceownerunrankedraw
```

BigEasyJ reviewed (approved):
Looks good, thanks for all of the work and collaboration!
